ext/sockets/sockets.c

```diff
@@ -1871,7 +1871,11 @@ PHP_FUNCTION(socket_set_option)
 			const char l_onoff_key[] = "l_onoff";
 			const char l_linger_key[] = "l_linger";
 
-			convert_to_array(arg4);
+			if (Z_TYPE_P(arg4) != IS_ARRAY) {
+				zend_argument_type_error(4, "must be of type array, %s given", zend_zval_value_name(arg4));
```

Comment:
I think adding what option we are talking about would help the error message.
```suggestion
				zend_argument_type_error(4, "must be of type array when argument #3 ($option) is SO_LINGER, %s given", zend_zval_value_name(arg4));
```

And ditto for the other error cases.